### PR TITLE
expression: Implement `VecEvalBool` to prepare for vectorized selection operator

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve? 
Implement `VecEvalBool`, which will be used in the `Selection` executor in the next PR.

### What is changed and how it works?
All types gain some performance benefits, and here is the benchmark:
```
BenchmarkVecEvalBool/Vec-int/-12                  100000             17659 ns/op              40 B/op          2 allocs/op
BenchmarkVecEvalBool/Row-int/-12                  100000             27787 ns/op               0 B/op          0 allocs/op
BenchmarkVecEvalBool/Vec-real/-12                  50000             24957 ns/op              40 B/op          2 allocs/op
BenchmarkVecEvalBool/Row-real/-12                  50000             30823 ns/op               0 B/op          0 allocs/op
BenchmarkVecEvalBool/Vec-decimal/-12               10000            192854 ns/op           60023 B/op       1914 allocs/op
BenchmarkVecEvalBool/Row-decimal/-12               10000            222921 ns/op           59904 B/op       1912 allocs/op
BenchmarkVecEvalBool/Vec-string/-12                20000             62754 ns/op              32 B/op          1 allocs/op
BenchmarkVecEvalBool/Row-string/-12                20000             69714 ns/op               0 B/op          0 allocs/op
BenchmarkVecEvalBool/Vec-datetime/-12              20000             74721 ns/op           30837 B/op        963 allocs/op
BenchmarkVecEvalBool/Row-datetime/-12              20000             83778 ns/op           30752 B/op        961 allocs/op
BenchmarkVecEvalBool/Vec-timestamp/-12             20000             77877 ns/op           31159 B/op        973 allocs/op
BenchmarkVecEvalBool/Row-timestamp/-12             20000             83610 ns/op           31072 B/op        971 allocs/op
BenchmarkVecEvalBool/Vec-duration/-12             100000             18779 ns/op              40 B/op          2 allocs/op
BenchmarkVecEvalBool/Row-duration/-12              50000             24744 ns/op               0 B/op          0 allocs/op
BenchmarkVecEvalBool/Vec-int/int/-12               50000             33948 ns/op              48 B/op          3 allocs/op
BenchmarkVecEvalBool/Row-int/int/-12               30000             45867 ns/op               0 B/op          0 allocs/op
BenchmarkVecEvalBool/Vec-real/int/-12              50000             28956 ns/op              48 B/op          3 allocs/op
BenchmarkVecEvalBool/Row-real/int/-12              50000             40628 ns/op               0 B/op          0 allocs/op
BenchmarkVecEvalBool/Vec-decimal/int/-12           10000            216174 ns/op           61331 B/op       1961 allocs/op
BenchmarkVecEvalBool/Row-decimal/int/-12            5000            245236 ns/op           61200 B/op       1958 allocs/op
BenchmarkVecEvalBool/Vec-string/int/-12            20000             76811 ns/op              40 B/op          2 allocs/op
BenchmarkVecEvalBool/Row-string/int/-12            20000             97678 ns/op               0 B/op          0 allocs/op
BenchmarkVecEvalBool/Vec-datetime/int/-12                  20000             90148 ns/op           30847 B/op        964 allocs/op
BenchmarkVecEvalBool/Row-datetime/int/-12                  10000            101904 ns/op           30752 B/op        961 allocs/op
BenchmarkVecEvalBool/Vec-timestamp/int/-12                 20000             89085 ns/op           31296 B/op        978 allocs/op
BenchmarkVecEvalBool/Row-timestamp/int/-12                 10000            102121 ns/op           31200 B/op        975 allocs/op
BenchmarkVecEvalBool/Vec-duration/int/-12                  50000             31702 ns/op              48 B/op          3 allocs/op
BenchmarkVecEvalBool/Row-duration/int/-12                  30000             38985 ns/op               0 B/op          0 allocs/op
BenchmarkVecEvalBool/Vec-int/real/-12                      50000             35173 ns/op              48 B/op          3 allocs/op
BenchmarkVecEvalBool/Row-int/real/-12                      30000             46434 ns/op               0 B/op          0 allocs/op
BenchmarkVecEvalBool/Vec-real/real/-12                     50000             29185 ns/op              48 B/op          3 allocs/op
BenchmarkVecEvalBool/Row-real/real/-12                     30000             41056 ns/op               0 B/op          0 allocs/op
BenchmarkVecEvalBool/Vec-decimal/real/-12                  10000            217434 ns/op           61569 B/op       1967 allocs/op
BenchmarkVecEvalBool/Row-decimal/real/-12                  10000            231550 ns/op           61440 B/op       1964 allocs/op
BenchmarkVecEvalBool/Vec-string/real/-12                   20000             79319 ns/op              40 B/op          2 allocs/op
BenchmarkVecEvalBool/Row-string/real/-12                   20000             97202 ns/op               0 B/op          0 allocs/op
BenchmarkVecEvalBool/Vec-datetime/real/-12                 20000             91679 ns/op           30845 B/op        964 allocs/op
BenchmarkVecEvalBool/Row-datetime/real/-12                 10000            108775 ns/op           30752 B/op        961 allocs/op
BenchmarkVecEvalBool/Vec-timestamp/real/-12                20000             91766 ns/op           31263 B/op        977 allocs/op
BenchmarkVecEvalBool/Row-timestamp/real/-12                10000            108107 ns/op           31168 B/op        974 allocs/op
BenchmarkVecEvalBool/Vec-duration/real/-12                 50000             34734 ns/op              48 B/op          3 allocs/op
BenchmarkVecEvalBool/Row-duration/real/-12                 30000             48356 ns/op               0 B/op          0 allocs/op
BenchmarkVecEvalBool/Vec-int/decimal/-12                   10000            202483 ns/op           58175 B/op       1861 allocs/op
BenchmarkVecEvalBool/Row-int/decimal/-12                   10000            213444 ns/op           58048 B/op       1858 allocs/op
BenchmarkVecEvalBool/Vec-real/decimal/-12                  10000            110048 ns/op           28236 B/op        899 allocs/op
BenchmarkVecEvalBool/Row-real/decimal/-12                  10000            122885 ns/op           28128 B/op        896 allocs/op
BenchmarkVecEvalBool/Vec-decimal/decimal/-12                5000            375046 ns/op          117051 B/op       3733 allocs/op
BenchmarkVecEvalBool/Row-decimal/decimal/-12                3000            394787 ns/op          116848 B/op       3730 allocs/op
BenchmarkVecEvalBool/Vec-string/decimal/-12                10000            231483 ns/op           54711 B/op       1744 allocs/op
BenchmarkVecEvalBool/Row-string/decimal/-12                 5000            272048 ns/op           54592 B/op       1742 allocs/op
BenchmarkVecEvalBool/Vec-datetime/decimal/-12               5000            267422 ns/op           89536 B/op       2835 allocs/op
BenchmarkVecEvalBool/Row-datetime/decimal/-12               5000            275322 ns/op           89360 B/op       2832 allocs/op
BenchmarkVecEvalBool/Vec-timestamp/decimal/-12              5000            265839 ns/op           90058 B/op       2847 allocs/op
BenchmarkVecEvalBool/Row-timestamp/decimal/-12              5000            277241 ns/op           89888 B/op       2844 allocs/op
BenchmarkVecEvalBool/Vec-duration/decimal/-12              10000            202091 ns/op           58029 B/op       1853 allocs/op
BenchmarkVecEvalBool/Row-duration/decimal/-12              10000            212624 ns/op           57904 B/op       1850 allocs/op
BenchmarkVecEvalBool/Vec-int/string/-12                    20000             71532 ns/op              40 B/op          2 allocs/op
BenchmarkVecEvalBool/Row-int/string/-12                    20000             83353 ns/op               0 B/op          0 allocs/op
BenchmarkVecEvalBool/Vec-real/string/-12                   30000             46312 ns/op              40 B/op          2 allocs/op
BenchmarkVecEvalBool/Row-real/string/-12                   30000             59413 ns/op               0 B/op          0 allocs/op
BenchmarkVecEvalBool/Vec-decimal/string/-12                 5000            252936 ns/op           60775 B/op       1936 allocs/op
BenchmarkVecEvalBool/Row-decimal/string/-12                 5000            295207 ns/op           60656 B/op       1934 allocs/op
BenchmarkVecEvalBool/Vec-string/string/-12                 10000            109359 ns/op              32 B/op          1 allocs/op
BenchmarkVecEvalBool/Row-string/string/-12                 10000            122881 ns/op               0 B/op          0 allocs/op
BenchmarkVecEvalBool/Vec-datetime/string/-12               10000            130041 ns/op           31318 B/op        978 allocs/op
BenchmarkVecEvalBool/Row-datetime/string/-12               10000            154637 ns/op           31232 B/op        976 allocs/op
BenchmarkVecEvalBool/Vec-timestamp/string/-12              10000            126836 ns/op           30773 B/op        961 allocs/op
BenchmarkVecEvalBool/Row-timestamp/string/-12              10000            143604 ns/op           30688 B/op        959 allocs/op
BenchmarkVecEvalBool/Vec-duration/string/-12               20000             72588 ns/op              40 B/op          2 allocs/op
BenchmarkVecEvalBool/Row-duration/string/-12               20000             85658 ns/op               0 B/op          0 allocs/op
BenchmarkVecEvalBool/Vec-int/datetime/-12                  20000             83778 ns/op           28922 B/op        904 allocs/op
BenchmarkVecEvalBool/Row-int/datetime/-12                  20000             99168 ns/op           28832 B/op        901 allocs/op
BenchmarkVecEvalBool/Vec-real/datetime/-12                 30000             54483 ns/op           14741 B/op        461 allocs/op
BenchmarkVecEvalBool/Row-real/datetime/-12                 20000             71652 ns/op           14656 B/op        458 allocs/op
BenchmarkVecEvalBool/Vec-decimal/datetime/-12               5000            262667 ns/op           89771 B/op       2842 allocs/op
BenchmarkVecEvalBool/Row-decimal/datetime/-12               5000            280161 ns/op           89600 B/op       2839 allocs/op
BenchmarkVecEvalBool/Vec-string/datetime/-12               10000            126326 ns/op           28020 B/op        875 allocs/op
BenchmarkVecEvalBool/Row-string/datetime/-12               10000            145406 ns/op           27936 B/op        873 allocs/op
BenchmarkVecEvalBool/Vec-datetime/datetime/-12             10000            142001 ns/op           60171 B/op       1879 allocs/op
BenchmarkVecEvalBool/Row-datetime/datetime/-12             10000            156342 ns/op           60032 B/op       1876 allocs/op
BenchmarkVecEvalBool/Vec-timestamp/datetime/-12            10000            142869 ns/op           60105 B/op       1877 allocs/op
BenchmarkVecEvalBool/Row-timestamp/datetime/-12            10000            154880 ns/op           59968 B/op       1874 allocs/op
BenchmarkVecEvalBool/Vec-duration/datetime/-12             20000             86628 ns/op           29883 B/op        934 allocs/op
BenchmarkVecEvalBool/Row-duration/datetime/-12             20000            103630 ns/op           29792 B/op        931 allocs/op
BenchmarkVecEvalBool/Vec-int/timestamp/-12                 20000             86872 ns/op           29595 B/op        925 allocs/op
BenchmarkVecEvalBool/Row-int/timestamp/-12                 10000            102888 ns/op           29504 B/op        922 allocs/op
BenchmarkVecEvalBool/Vec-real/timestamp/-12                30000             56253 ns/op           15509 B/op        485 allocs/op
BenchmarkVecEvalBool/Row-real/timestamp/-12                20000             69588 ns/op           15424 B/op        482 allocs/op
BenchmarkVecEvalBool/Vec-decimal/timestamp/-12              5000            264613 ns/op           91132 B/op       2888 allocs/op
BenchmarkVecEvalBool/Row-decimal/timestamp/-12              5000            276208 ns/op           90960 B/op       2885 allocs/op
BenchmarkVecEvalBool/Vec-string/timestamp/-12              10000            126208 ns/op           27537 B/op        860 allocs/op
BenchmarkVecEvalBool/Row-string/timestamp/-12              10000            140041 ns/op           27456 B/op        858 allocs/op
BenchmarkVecEvalBool/Vec-datetime/timestamp/-12            10000            144482 ns/op           60423 B/op       1887 allocs/op
BenchmarkVecEvalBool/Row-datetime/timestamp/-12            10000            161203 ns/op           60288 B/op       1884 allocs/op
BenchmarkVecEvalBool/Vec-timestamp/timestamp/-12           10000            142248 ns/op           60903 B/op       1902 allocs/op
BenchmarkVecEvalBool/Row-timestamp/timestamp/-12           10000            158567 ns/op           60768 B/op       1899 allocs/op
BenchmarkVecEvalBool/Vec-duration/timestamp/-12            20000             85299 ns/op           29723 B/op        929 allocs/op
BenchmarkVecEvalBool/Row-duration/timestamp/-12            10000            104636 ns/op           29632 B/op        926 allocs/op
BenchmarkVecEvalBool/Vec-int/duration/-12                  50000             31097 ns/op              48 B/op          3 allocs/op
BenchmarkVecEvalBool/Row-int/duration/-12                  50000             40184 ns/op               0 B/op          0 allocs/op
BenchmarkVecEvalBool/Vec-real/duration/-12                 50000             28523 ns/op              48 B/op          3 allocs/op
BenchmarkVecEvalBool/Row-real/duration/-12                 30000             42096 ns/op               0 B/op          0 allocs/op
BenchmarkVecEvalBool/Vec-decimal/duration/-12              10000            210339 ns/op           60352 B/op       1933 allocs/op
BenchmarkVecEvalBool/Row-decimal/duration/-12              10000            223314 ns/op           60224 B/op       1930 allocs/op
BenchmarkVecEvalBool/Vec-string/duration/-12               20000             72792 ns/op              40 B/op          2 allocs/op
BenchmarkVecEvalBool/Row-string/duration/-12               20000             91357 ns/op               0 B/op          0 allocs/op
BenchmarkVecEvalBool/Vec-datetime/duration/-12             20000             89759 ns/op           31166 B/op        974 allocs/op
BenchmarkVecEvalBool/Row-datetime/duration/-12             10000            100507 ns/op           31072 B/op        971 allocs/op
BenchmarkVecEvalBool/Vec-timestamp/duration/-12            20000             88644 ns/op           31199 B/op        975 allocs/op
BenchmarkVecEvalBool/Row-timestamp/duration/-12            10000            103003 ns/op           31104 B/op        972 allocs/op
BenchmarkVecEvalBool/Vec-duration/duration/-12             50000             31221 ns/op              48 B/op          3 allocs/op
BenchmarkVecEvalBool/Row-duration/duration/-12             50000             42148 ns/op               0 B/op          0 allocs/op
```

### Check List 

Tests 

 - Unit test

### CODE REVIEW

File: expression/expression.go [hunk 1]
```diff
@@ -208,6 +209,125 @@ func EvalBool(ctx sessionctx.Context, exprList CNFExprs, row chunk.Row) (bool, b
...
+					d.SetString(buf.GetString(i))
+				case types.ETJson:
+					d.SetMysqlJSON(buf.GetJSON(i))
+				case types.ETDecimal:
+					d.SetMysqlDecimal(buf.GetDecimal(i))
+				}
+			}
+
+			if d.IsNull() {
+				if !isEQCondFromIn { // same as EvalBool
```
Comment (jingyugao) [inline expression/expression.go]: how about move this code into `if buf.IsNull(i) {`
or use `SetNull` replace `SetValue(nil)`

Comment (qw4990) [reply expression/expression.go]: updated

File: expression/expression.go [hunk 2]
```diff
@@ -208,6 +209,125 @@ func EvalBool(ctx sessionctx.Context, exprList CNFExprs, row chunk.Row) (bool, b
...
+func VecEvalBool(ctx sessionctx.Context, exprList CNFExprs, input *chunk.Chunk, selected, nulls []bool) ([]bool, []bool, error) {
+	n := input.NumRows()
+	selected = selected[:0]
+	nulls = nulls[:0]
+	for i := 0; i < n; i++ {
+		selected = append(selected, false)
+		nulls = append(nulls, false)
+	}
+
+	defer input.SetSel(input.Sel())
```
Comment (XuHuaiyu) [inline expression/expression.go]: Can this line be moved to line 237

File: expression/expression.go [hunk 3]
```diff
@@ -208,6 +209,125 @@ func EvalBool(ctx sessionctx.Context, exprList CNFExprs, row chunk.Row) (bool, b
...
+			return nil, nil, err
+		}
+
+		isEQCondFromIn := IsEQCondFromIn(expr)
+		hasUnsignedFlag := mysql.HasUnsignedFlag(expr.GetType().Flag)
+		d, j := types.Datum{}, 0
+		for i := range sel {
+			if buf.IsNull(i) {
+				if !isEQCondFromIn {
+					selected[sel[i]] = false
```
Comment (XuHuaiyu) [inline expression/expression.go]: Is this line useless?

File: expression/expression.go [hunk 4]
```diff
@@ -208,6 +209,125 @@ func EvalBool(ctx sessionctx.Context, exprList CNFExprs, row chunk.Row) (bool, b
...
+	}
+
+	defer input.SetSel(input.Sel())
+	sel := selPool.Get().([]int)
+	defer selPool.Put(sel)
+	sel = sel[:0]
+	for i := 0; i < n; i++ {
+		sel = append(sel, i)
+	}
+	input.SetSel(sel)
```
Comment (XuHuaiyu) [inline expression/expression.go]: Will it be promised that there is no gap in input.sel?

File: expression/expression.go [hunk 5]
```diff
@@ -208,6 +209,125 @@ func EvalBool(ctx sessionctx.Context, exprList CNFExprs, row chunk.Row) (bool, b
...
+var (
+	selPool = sync.Pool{
+		New: func() interface{} {
+			return make([]int, 1024)
+		},
+	}
+)
+
+// VecEvalBool does the same thing as EvalBool but it works in a vectorized manner.
+func VecEvalBool(ctx sessionctx.Context, exprList CNFExprs, input *chunk.Chunk, selected, nulls []bool) ([]bool, []bool, error) {
```
Comment (XuHuaiyu) [inline expression/expression.go]: Will it be promised that there is no gap in input.sel before this line?

File: expression/expression.go (hunk 3 above)
Comment (qw4990) [reply expression/expression.go]: You are right.

File: expression/expression.go (hunk 5 above)
Comment (qw4990) [reply expression/expression.go]: It is premised now, but we may break this promise later for further optimization.

File: expression/expression.go (hunk 2 above)
Comment (qw4990) [reply expression/expression.go]: updated

File: expression/expression.go (hunk 3 above)
Comment (qw4990) [reply expression/expression.go]: This line has been removed.

**Pull request merged**